packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -1,8 +1,72 @@
 pragma solidity ^0.6.0;
 pragma experimental ABIEncoderV2;
 
+import "../interfaces/FundingRateStoreInterface.sol";
+import "../../../common/implementation/Testable.sol";
 
-// TODO:
-contract FundingRateStore {
 
+contract FundingRateStore is FundingRateStoreInterface, Testable {
+    struct Proposal {
+        FixedPoint.Signed rate;
+        uint256 time;
+        address proposer;
+    }
+
+    struct FundingRateRecord {
+        FixedPoint.Signed rate;
+        Proposal proposal;
+    }
+
+    enum ProposalState { None, Pending, Expired }
+
+    mapping(bytes32 => FundingRateRecord) public fundingRateRecords;
+
+    uint256 public proposalLiveness;
+
+    constructor(uint256 _proposalLiveness, address _timerAddress) public Testable(_timerAddress) {
+        require(_proposalLiveness > 0, "Proposal liveness is 0");
+        proposalLiveness = _proposalLiveness;
+    }
+
+    function getFundingRateForIdentifier(bytes32 identifier) external override view returns (FixedPoint.Signed memory) {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+
+        if (_getProposalState(fundingRateRecord.proposal) == ProposalState.Expired) {
+            return fundingRateRecord.proposal.rate;
+        } else {
+            return fundingRateRecord.rate;
+        }
+    }
+
+    function propose(bytes32 identifier, FixedPoint.Signed memory rate) external {
+        // TODO: ACLS on identifiers that can be proposed.
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(identifier);
+
+        ProposalState proposalState = _getProposalState(fundingRateRecord.proposal);
+
+        // TODO: bond logic.
+
+        require(proposalState != ProposalState.Pending, "Existing proposal still pending.");
+
+        if (proposalState == ProposalState.Expired) {
+            fundingRateRecord.rate = fundingRateRecord.proposal.rate;
+        }
+
+        fundingRateRecord.proposal = Proposal({ rate: rate, time: getCurrentTime(), proposer: msg.sender });
```

Comment:
As described in #2061 I think `propose()` should also revert if `rate == fundingRateRecord.rate` (proposer is proposing the same rate as the current). This prevents empty updates.

Reply:
Does this create a strange incentive to push a funding rate that's _slightly_ different just to get the reward? Put another way, is there any problem with someone pushing an update that says "the funding rate is the same"?

Reply:
I guess this depends on how rewards are structured, imagine 2 different reward structures:

- Rewards are proportional to delta between proposed rate and current rate:
    - Proposers have no reason to push an update that says "the funding rate is the same"
    - Proposers are incentivized to push funding rates IFF the rate has deviated above a certain profitable threshold
- Rewards reset to 0 after each update, grow linearly:
    - Proposers are incentivized to push an update that says "the funding rate is the same", unless the Store explicitly prevents these updates
    - Proposers ignore how much the proposed rate has deviated from the current rate

Reply:
Yes, totally agree.

I think, rather than reverting, we should just build this into the incentivization structure. However, like you say below, I think this is a q for a later PR. Thanks for bringing it up!